### PR TITLE
#8 LCOM4: initial implementation

### PR DESCRIPTION
**Pull request by llorllale**
as per #8 

> 'LCOM4' is a 1995 revision by Martin Hitz and Behzad Montazeri,
          that further improves upon the revised versions LCOM3 proposed by
          W. Li and S. Henry in "Maintenance Metrics for the Object Oriented Paradigm".
          LCOM4 establishes connections between methods not just on shared attributes
          of the class, but also on whether one method calls the other. That is,
          methods A and B are said to be related if either both use at least one
          instance variable in common, OR A calls B, OR B calls A.
          LCOM4 tweaks the formula for method connections in order to eliminate
          the distorting effects that 'getters' may have on earlier LCOM variants,
          and also methods that purely rely on the reuse of other methods without
          accessing attributes directly (think method overloads where the ones
          with the least args reuse the ones with the most args).

This PR:
* Updates README
* Includes the relevant PDF
* Provides initial implementation of LCOM4 along with a testcase
* Adds new test case `MethodMethodCalls`
* Leaves puzzle for implementing the rest of the test cases
* Leaves puzzle due to an impediment

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#206](https://github.com/yegor256/jpeek/issues/206) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#206](https://github.com/yegor256/jpeek/issues/206) is assigned to @amihaiemil[/z](https://www.0crat.com/u/amihaiemil), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/502049086). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by llorllale** [comment]
@amihaiemil please review

**Comment by llorllale** [comment]
ping @amihaiemil

File: README.md [hunk 1]
```diff
@@ -166,6 +166,13 @@ Cohesion and Reuse in an Object-Oriented System,<br/>
 Department of Computer Science, Colorado State University, 1995,
 [PDF](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.53.2683).
 
+[`hitz95`]
+Lack of Cohesion in Methods (**LCOM4**).<br/>
```
Comment (amihaiemil) [inline README.md]: @llorllale "Lack of Cohesion in Methods 4 (**LCOM4**)"

File: src/main/resources/org/jpeek/metrics/LCOM4.xsl [hunk 2]
```diff
@@ -0,0 +1,110 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (amihaiemil) [inline src/main/resources/org/jpeek/metrics/LCOM4.xsl]: @llorllale Is the PDF necessary, should it be in the repo? Wouldn't a link to it in the README be better?

File: src/main/resources/org/jpeek/metrics/LCOM4.xsl [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    </metric>
+  </xsl:template>
+  <xsl:template match="class">
+    <xsl:variable name="class_fqn" select="replace(string-join(../@id | @id, '.'), '^\.', '')"/>
+    <xsl:variable name="A" select="attributes/attribute[@static = 'false']/text()"/>
+    <xsl:variable name="a" select="count($A)"/>
+    <xsl:variable name="M" select="methods/method"/>
+    <xsl:variable name="m" select="count($M)"/>
+    <!--
+    @todo #8:30min LCOM4: #156 needs to be fixed in order to avoid mistakenly capturing ops
```
Comment (amihaiemil) [inline src/main/resources/org/jpeek/metrics/LCOM4.xsl]: @llorllale I don't really understand this todo: "to avoid mistakenly capturing ops on fields as if they actually belong to our enclosing class instead of another class" -- which is the "enclosing" class and which is "another" class? Maybe rephrase it?

File: README.md (hunk 1 above)
Comment (llorllale) [reply README.md]: @amihaiemil ok

File: src/main/resources/org/jpeek/metrics/LCOM4.xsl (hunk 2 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/metrics/LCOM4.xsl]: @amihaiemil I don't know why but we're always adding the PDFs to the repo. I do agree that we shouldn't be committing these binaries. Perhaps just attaching them to the ticket should suffice. Not my call.

File: src/main/resources/org/jpeek/metrics/LCOM4.xsl (hunk 3 above)
Comment (llorllale) [reply src/main/resources/org/jpeek/metrics/LCOM4.xsl]: @amihaiemil ok

**Comment by llorllale** [comment]
@amihaiemil I fixed the typo in README and clarified the puzzle a bit. With regards to the PDF, we've always done it that way.

Please review

**Comment by amihaiemil** [comment]
@rultor good to merge

**Comment by llorllale** [comment]
@yegor256 ping

**Comment by 0crat** [comment]
@amihaiemil[/z](https://www.0crat.com/u/amihaiemil) this job was assigned to you 5 days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by amihaiemil** [comment]
@yegor256 can we merge this?

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job, as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by ypshenychka** [comment]
@amihaiemil According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only one cosmetic issue was found during code review. 
Please confirm that you'll try to find at least major three problems while future reviews.

**Comment by amihaiemil** [comment]
@ypshenychka I confirm :D

**Comment by ypshenychka** [comment]
@amihaiemil Thank you!

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
`QA` review completed: +8 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka), total is [+742](https://www.0crat.com/u/ypshenychka)

**Comment by 0crat** [comment]
> @0crat quality acceptable ([here](https://github.com/yegor256/jpeek/issues/206#issuecomment-368575647))

@ypshenychka You can't do that, unless you have one of these roles: `ARC`, `PO`. Your current roles are: `QA`.

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256), total is [+10939](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#206](https://github.com/yegor256/jpeek/issues/206) is now out of scope
